src/common/acl_principal.c

```diff
@@ -0,0 +1,307 @@
+ * function.
+ */
+#define TRY_UNTIL_BUF_SIZE_OK(expr)				\
+do {								\
+	char	*new_buf = NULL;				\
+	size_t	buflen;						\
+								\
+	buflen = DEFAULT_BUF_LEN;				\
+	do {							\
+		D_REALLOC(new_buf, buf, buflen);		\
+		if (new_buf == NULL)				\
+			D_GOTO(out, rc = -DER_NOMEM);		\
+		buf = new_buf;					\
+								\
+		rc = (expr);					\
+								\
+		buflen *= 2;					\
+	} while (rc == ERANGE);					\
+} while (0)
+
+/*
+ * States used to parse a principal name
+ */
+enum validity_state {
+	STATE_START,
+	STATE_NAME,
+	STATE_DOMAIN,
+	STATE_DONE,
+	STATE_INVALID
+};
+
+static enum validity_state
+next_validity_state(enum validity_state current_state, char ch)
+{
+	switch (current_state) {
+	case STATE_START:
+		if (ch == '@')
+			return STATE_INVALID;
+		return STATE_NAME;
+	case STATE_NAME:
+		if (ch == '@')
+			return STATE_DOMAIN;
+		if (ch == '\0')
+			return STATE_INVALID;
+		break;
+	case STATE_DOMAIN:
+		if (ch == '\0')
+			return STATE_DONE;
+		if (ch == '@')
+			return STATE_INVALID;
+		break;
+	case STATE_DONE:
+	case STATE_INVALID:
+		break;
+	default:
+		D_ASSERTF(false, "Invalid state: %d\n", current_state);
+	}
+
+	return current_state;
+}
+
+bool
+daos_acl_principal_is_valid(const char *name)
+{
+	size_t			len;
+	size_t			i;
+	enum validity_state	state = STATE_START;
+
+	if (name == NULL) {
+		D_INFO("Name was NULL\n");
+		return false;
+	}
+
+	len = strnlen(name, DAOS_ACL_MAX_PRINCIPAL_BUF_LEN);
+	if (len == 0 || len > DAOS_ACL_MAX_PRINCIPAL_LEN) {
+		D_INFO("Invalid len: %lu\n", len);
+		return false;
+	}
+
+	for (i = 0; i < (len + 1); i++) {
+		state = next_validity_state(state, name[i]);
+		if (state == STATE_INVALID) {
+			D_INFO("Name was badly formatted: %s\n", name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int
+local_name_to_principal_name(const char *local_name, char **name)
+{
+	D_ASPRINTF(*name, "%s@", local_name);
+	if (*name == NULL) {
+		D_ERROR("Failed to allocate string for name\n");
+		return -DER_NOMEM;
+	}
+
+	return 0;
+}
+
+int
+daos_acl_uid_to_principal(uid_t uid, char **name)
+{
+	int		rc;
+	struct passwd	user;
+	struct passwd	*result = NULL;
+	char		*buf = NULL;
+
+	if (name == NULL) {
+		D_INFO("name pointer was NULL!\n");
+		return -DER_INVAL;
+	}
+
+	TRY_UNTIL_BUF_SIZE_OK(getpwuid_r(uid, &user, buf, buflen, &result));
+
+	if (rc != 0) {
+		D_ERROR("Error from getpwuid_r: %d\n", rc);
+		D_GOTO(out, rc = d_errno2der(rc));
+	}
+
+	if (result == NULL) {
+		D_INFO("No user for uid %u\n", uid);
+		D_GOTO(out, rc = -DER_NONEXIST);
+	}
+
+	rc = local_name_to_principal_name(result->pw_name, name);
+
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+int
+daos_acl_gid_to_principal(gid_t gid, char **name)
+{
+	int		rc;
+	struct group	grp;
+	struct group	*result = NULL;
+	char		*buf = NULL;
+
+	if (name == NULL) {
+		D_INFO("name pointer was NULL!\n");
+		return -DER_INVAL;
+	}
+
+	TRY_UNTIL_BUF_SIZE_OK(getgrgid_r(gid, &grp, buf, buflen, &result));
```

Comment:
I think it would be a little cleaner if we did something like this since each expression takes same number of parameters.  

rc = TRY..(func, arg1, arg2, buf, &result);  passing buflen is a little weird.

You can make a macro return a value by wrapping it with 

({
     int __rc = 0;
     /* rest of macro */
     __rc;   // last statement is "return" value
})

Reply:
Good suggestion, that takes some of the weird assumptions out.